python/ray/data/_internal/stats.py

```diff
@@ -275,137 +332,302 @@ def summary_string(
                 stage_uuid = self.dataset_uuid + stage_name
                 out += "\n"
                 out += "\tSubstage {} {}: ".format(n, stage_name)
                 if stage_uuid in already_printed:
                     out += "\t[execution cached]\n"
                 else:
                     already_printed.add(stage_uuid)
-                    out += self._summarize_blocks(metadata, is_substage=True)
-        out += self._summarize_iter()
+                    out += str(stage_stats_summary)
+        out += str(self.iter_stats)
+        if self.extra_metrics:
+            indent = "\t" if stage_stats_summary.is_substage else ""
+            out += indent
+            out += "* Extra metrics: " + str(self.extra_metrics) + "\n"
         return out
 
-    def _summarize_iter(self) -> str:
-        out = ""
-        if (
-            self.iter_total_s.get()
-            or self.iter_wait_s.get()
-            or self.iter_next_batch_s.get()
-            or self.iter_format_batch_s.get()
-            or self.iter_get_s.get()
-        ):
-            out += "\nDataset iterator time breakdown:\n"
-            out += "* In ray.wait(): {}\n".format(fmt(self.iter_wait_s.get()))
-            out += "* In ray.get(): {}\n".format(fmt(self.iter_get_s.get()))
-            out += "* In next_batch(): {}\n".format(fmt(self.iter_next_batch_s.get()))
-            out += "* In format_batch(): {}\n".format(
-                fmt(self.iter_format_batch_s.get())
-            )
-            out += "* In user code: {}\n".format(fmt(self.iter_user_s.get()))
-            out += "* Total time: {}\n".format(fmt(self.iter_total_s.get()))
-        return out
+    def get_total_wall_time(self) -> float:
+        parent_sum = sum(p.get_total_wall_time() for p in self.parents)
+        return parent_sum + sum(
+            getattr(ss, StageStatsSummary.WALL_TIME).get("max", 0)
+            for ss in self.stages_stats
+        )
+
+    def get_total_cpu_time(self) -> float:
+        parent_sum = sum(p.get_total_cpu_time() for p in self.parents)
+        return parent_sum + sum(
+            getattr(ss, StageStatsSummary.CPU_TIME).get("sum", 0)
+            for ss in self.stages_stats
+        )
+
+    def get_max_heap_memory(self) -> float:
+        parent_memory = [p.get_max_heap_memory() for p in self.parents]
+        parent_max = max(parent_memory) if parent_memory else 0
+        return max(
+            parent_max,
+            *[
+                getattr(ss, StageStatsSummary.PEAK_HEAP_MEMORY).get("max", 0)
+                for ss in self.stages_stats
+            ],
+        )
+
+
+@dataclass
+class StageStatsSummary:
+    # Keys used in initialization params
+    BLOCK_EXECUTION_SUMMARY = "block_execution_summary_str"
+    WALL_TIME = "wall_time"
+    CPU_TIME = "cpu_time"
+    PEAK_HEAP_MEMORY = "memory"
+    OUTPUT_NUM_ROWS = "output_num_rows"
+    OUTPUT_SIZE_BYTES = "output_size_bytes"
+    NODE_COUNT = "node_count"
```

Comment:
I don't think that we need these keys now that they're defined as attributes on the dataclass, since the attributes themselves are just as discoverable as the keys mapping to the attribute names so this extra layer of indirection doesn't seem to give us much.

Instead of rolling up a kwarg dictionary in `.from_block_metadata()`, we could have conditional definition code for the optional attributes like memory:
```python
if exec_stats:
    ....
    memory = {...}
else:
    memory = None
```

This would also have the added benefit of eliminating the `getattr()` calls in `StageStatsSummary.__str__` and in the `DatasetStatsSummary.get_*()` methods, since the attributes can be accessed directly.